### PR TITLE
[WIP] monitoring super container

### LINKED ISSUE #1274: Prometheus/Grafana bundled supercontainer
## Abstract
Let's move our Prometheus/Grafana tools into one Docker container that can be configured with environment variables.

## Motivation
Recently, there have been a few requests for go-livepeer monitoring in a variety of contexts:

* We have a variety of Kubernetes deployments that require monitoring. Even within the same cluster, we sometimes want different deployments to be monitored separately.
* Folks are running [the docker-compose BOT setup](https://github.com/mk-livepeer/bot-miner-setup) and would like stats
* Many users are just running go-livepeer binaries and would like stats there too.

Our existing Prometheus deployments within the test-harness or the Kubernetes clusters are specialized for their environments and difficult to dynamically adapt to these different contexts. [Prometheus does not support configuration with environment variables](https://github.com/prometheus/prometheus/issues/2357), further complicating dynamic configuration.

## Proposed Solution

Let's combine our Prometheus and Grafana configurations into a single Docker container. This container will be configured by environment variables. At runtime, a script in the container will take care of generating a prometheus.yml file appropriate for the environment. Here are some examples:

For a host that's just running two go-livepeer binaries, say

`livepeer -broadcaster -metrics -cliAddr=127.0.0.1:7935 -orchAddr 127.0.0.1:8936`
and
`livepeer -orchestrator -transcoder -serviceAddr 127.0.0.1:8936 -cliAddr=127.0.0.1:7936 -metrics=true`

The host could run a monitoring deployment that manually inspect these two targets

```
docker run
  -e LP_PROM_BROADCASTERS=http://127.0.0.1:7935/metrics
  -e LP_PROM_ORCHESTRATORS=http://127.0.0.1:7936/metrics
  livepeer/monitor
```

For a Kubernetes deployment, we could specify the value of the `prometheus.io/scrape` value to automatically discover and scrape from pertinent pods:

```
env:
- name: LP_PROM_KUBE_SCRAPE_VALUE
  value: testnet
```

wip the rest of this, need to hit a different meeting

### PR DESCRIPTION
**Pull request by ya7ya**
This container will include all the required services we need to run monitoring in various cloud providers and standalone setups.

fixes https://github.com/livepeer/go-livepeer/issues/1274

### CODE REVIEW

File: monitoring/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+             "$GF_PATHS_PLUGINS" \
+             "$GF_PATHS_DATA" && \
+    cp "$GF_PATHS_HOME/conf/sample.ini" "$GF_PATHS_CONFIG" && \
+    cp "$GF_PATHS_HOME/conf/ldap.toml" /etc/grafana/ldap.toml && \
+    chown -R grafana:grafana "$GF_PATHS_DATA" "$GF_PATHS_HOME/.aws" "$GF_PATHS_LOGS" "$GF_PATHS_PLUGINS" "$GF_PATHS_PROVISIONING" && \
+    chmod -R 777 "$GF_PATHS_DATA" "$GF_PATHS_HOME/.aws" "$GF_PATHS_LOGS" "$GF_PATHS_PLUGINS" "$GF_PATHS_PROVISIONING"
+
+# PhantomJS
+COPY --from=c1 /lib /lib
+COPY --from=c1 /lib64 /lib64
```
Comment (iameli) [inline monitoring/Dockerfile]: This kind of thing makes me nervous — are we going to step on any libraries added by any of the other dependencies? Are the naming conventions going to be consistent?

I'd be more inclined to start with a vanilla ubuntu or debian container and install Prometheus/Grafana/Phantom with a package manager — it's only a marginal increase in image size but is a lot easier to reason about.

File: monitoring/supervisord.conf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[supervisord]
+nodaemon=true
+
+[program:prometheus]
```
Comment (iameli) [inline monitoring/supervisord.conf]: Check this example for logging — the `/dev/fd/1` trick here seems to work well for getting stdout on the container. https://github.com/livepeer/livepeer-wowza/blob/master/etc/WowzaStreamingEngine.conf

**Review by iameli — CHANGES_REQUESTED** [review]

**Comment by ya7ya** [comment]
How to build the container for testing this PR
```bash
# in  docker-livepeer/monitoring 
$ sudo docker build -t super-container:latest -f Dockerfile .

# standalone mode listening to 2 livepeer nodes at 7935 and 7936 locally
$ sudo docker run --rm --name=monitoring --net=host --env LP_MODE=standalone --env LP_NODES=localhost:9735,localhost:7936 super-container:latest

# kubernetes with a namespace(s)
$ sudo docker run --rm --name=monitoring --net=host --env LP_MODE=kubernetes --env LP_KUBE_NAMESPACES=default,kube-system super-container:latest

```
 to see the generated `prometheus.yml`, simply `sudo docker exec -it monitoring cat /etc/prometheus/prometheus.yml` :+1:

**Review by iameli — CHANGES_REQUESTED** [review]
Looks solid! Just a few line-by-line things.

I have one suggestion that I think would make this a lot easier to use: let's handle environment variable/CLI parsing with `yargs`, such that the following two commands would work the same way:

```
docker run --env LP_MODE=standalone --env LP_NODES=localhost:9735,localhost:7936 livepeer/monitor
```

and

```
docker run livepeer/monitor --mode=standalone --nodes=localhost:9735,localhost:7936
```

I think the latter is easier to read and reason about. It also gives us a good place to put documentation; that's as easy as `docker run livepeer/monitor --help`, where we can explain all the usage information and caveats (`--net=host`) and whatnot.

I don't think this should be too hard to implement, you can copy the same pattern I used on the API here: https://github.com/livepeer/livepeerjs/blob/master/packages/api/src/cli.js#L21

Sorry for the last minute feature request! Totally okay to get this merged with the other changes first and open a new branch if you'd like to do it that way.

**Comment by iameli** [comment]
Would also be nice to get this into a CI pipeline so changes are pushed automatically. Take your pick, I think company-wide we're using like... six

File: monitoring/start.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+node /config-generator/src/generate.js
```
Comment (iameli) [inline monitoring/start.sh]: For yargs, we can pass in incoming CLI requests with... I think `node /config-generator/src/generate.js "$@"`. But maybe `$*` or `"$*"` or `$@` would work better, I can never keep track of those four

File: monitoring/config-generator/src/generate.js [hunk 4]
```diff
@@ -0,0 +1,364 @@
+'use strict'
+
+const fs = require('fs')
+const path = require('path')
+const YAML = require('yaml')
+
+function generate () {
+  console.log(process.env)
```
[the remark this answers is not in the document]
Comment (ya7ya) [reply monitoring/config-generator/src/generate.js]: yeah i'm gonna clean up all these console logs :+1:

**Review by iameli — APPROVED** [review]
LGTM! Let's write some documentation for `--help` and merge this during our 1-on-1 tomorrow.

**Pull request merged**